### ISSUE TITLE
Spike: Resource causality flow 3

### ISSUE THREAD

**Issue by clarafu** [concourse/concourse#7431]
## Summary

Working off of the last causality spike https://github.com/concourse/concourse/issues/6904, we now have a page where you can view the causality of a resource version. This means you can see all the builds and resource versions that are a result of a version, or you can see all the builds and resource versions that resulted in a version. Here are a few screenshots of these flows. 

This is the pipeline that I am using as an example.

<img width="1613" alt="Screen Shot 2021-08-17 at 4 39 37 PM" src="https://user-images.githubusercontent.com/19290703/129797517-57c593e9-d5c7-4893-9a7a-8084725b2084.png">

I can click on the `concourse` resource in order to see the resource version page for this resource. In that page, I can click on the `view all` button on the `inputs to` or `outputs of` section.

<img width="1498" alt="Screen Shot 2021-08-17 at 4 40 27 PM" src="https://user-images.githubusercontent.com/19290703/129797576-1c9fe9f5-0e31-412c-be5a-eb085d9047b8.png">

If I click on the `view all` button on the `inputs to` section, that will bring me to a causality view of all the downstream builds and resource versions that are a result of the version I clicked on.

<img width="1617" alt="Screen Shot 2021-08-17 at 4 40 47 PM" src="https://user-images.githubusercontent.com/19290703/129797803-04758557-f11b-4eba-bb8d-e4f556077023.png">

This causality page shows me a great amount of useful information if I am looking for the outputs that came out from a specific version of a resource. 

Using the initial problem statements, I find that this view is able to solve second problem statement very clearly. It is able to display the inputs/builds of upstream builds and versions and also the outputs/builds of downstream builds and versions in a clear and condensed fashion.

![Screen Shot 2021-07-23 at 11 07 06 AM](https://user-images.githubusercontent.com/19290703/129798099-ffa45d8c-675a-4819-bd20-28fa83f43d4e.png)

The next part of this causality epic will be solving the first problem statement, of knowing where a resource version is at in regards to a pipeline. This is very hard to figure out using the existing causality view because the placement of the jobs/builds  and resource/version nodes are not very predictable and placed very differently from the pipeline view, which most users are already comfortable with.

I think the best way to solve the first problem statement is to have some sort of extension of the existing pipeline view which will show "pipeline runs". These runs will show the progress of a single resource version through the pipeline. 

## Checklist Todo:
- [ ] API : Create Pipeline Causality API endpoint that will be used to fetch the data for loading up the initial pipeline causality view. This will return back the information for the "latest" run.
- [ ] Web : Create initial pipeline causality view in the web UI. This will be the view that will be shown when the user clicks to see the causality run of a version from the resource versions page. This view is similar to the pipeline view in terms of the nodes and edges, differences are that it will be showing the pipeline view with only the **latest** run using the version that the view is showing. It will also show the resource versions within the resource node that are associated with the causality version and also the build numbers within the job nodes that are associated with that causality version.
- [ ] API : Create Resource Version Causality API endpoint that will be used to fetch the versions that are associated to the version(s) used for the causality view
- [ ] API : Create Job Builds Causality API endpoint that will be used to fetch the builds that are associated to the version(s) used for the causality view
- [ ] Web : Add the button for a dropdown in the resource and job node in order to select a different version/build to show the run of. This will use the resource version causality api and the job builds causality api. Clicking a new version/build will refresh the Pipeline Causality API.
    - [ ] Spinner icon for when the web UI is waiting for an API response
- [ ] Web : Modify Pipeline Causality API endpoint to take in a build of a job in order to show the run for that specific build with the associated causality version(s)

Extra : 
- [ ] API : Modify the Pipeline Causality API to allow for multiple versions to be included in the causality view. It should send a request to the endpoint with the existing causality versions and the new version to show, then that should return back the causality state that should be shown in the web UI
- [ ] API : Modify the Pipeline Causality API to allow for multiple builds from different jobs to be included in the causality view. It should send a request to the endpoint with the fixed builds, then that should return back the causality state that should be shown in the web UI

## Things to Note

Currently, this view will only work for versions/builds run by the latest config. If the user were to fix a version or a build that was used in an older pipeline config, it will show up in the pipeline causality view using the latest pipeline config (probably will look jank). Not sure how to fix this, one idea that can possibly help is to use job pipes to construct the job/resource nodes but this won't show unused jobs/resources.

Will we be removing the existing causality view? Because this view can cover the problem statement that the existing causality view solves, which is what are the output versions produced by this input version.

## Context

* RFC: concourse/rfcs#
* Prior discussion: #
* Depends on #
* Part of #

**Comment by clarafu** [comment]
![Concourse causality - page 1](https://user-images.githubusercontent.com/19290703/130123379-05b93e96-cefc-4afd-9ff8-00d29329f3a3.png)
![Concourse causality - page 2](https://user-images.githubusercontent.com/19290703/130123394-fa083dcf-b4c8-40ad-bcd4-5e3eef49fd2c.png)
![Concourse causality - page 3](https://user-images.githubusercontent.com/19290703/130123401-d87f9e3e-17bd-40e8-a948-fdee7cb820c3.png)
![Concourse causality - page 4](https://user-images.githubusercontent.com/19290703/130123422-0bbc973b-7b20-44a8-9dcd-a9726444bcc8.png)
![Concourse causality - page 5](https://user-images.githubusercontent.com/19290703/130123452-6b77bd54-2740-433a-ac24-9288ac506e33.png)
![Concourse causality - page 6](https://user-images.githubusercontent.com/19290703/130123463-d04db2ad-0115-4ee1-a151-e8c07477e13e.png)
![Concourse causality - page 7](https://user-images.githubusercontent.com/19290703/130123472-fb50a40c-cbe2-45c5-8c65-605753f24c06.png)
![Concourse causality - page 8](https://user-images.githubusercontent.com/19290703/130123481-e1ee9342-5781-4fae-b7bf-74f340a13143.png)

**Comment by clarafu** [comment]
What kind of data do we need to support this new view?

In pipeline view :
- Resource versions when you click (while holding down a button) on the resource in the pipeline page in order to view the causality page.

In new causality view :
- The "run" that includes the pinned version(s)
    - the latest build of each job both directly or indirectly related to the pinned version(s)
    - the set of versions for each resource that was last used in the run
    - % of pass/fails of each job that uses the pinned version(s)
- Clicking on the job (while holding down button) should show a list of builds that are associated with the pinned version(s)

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
